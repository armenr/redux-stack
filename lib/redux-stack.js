'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildStack = undefined;

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _redux = require('redux');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

// TODO: stack formatter + printer dump at build time, etc.

// composition for arrays, hashes, and functions
var mergeStack = _ramda2.default.reduce(_ramda2.default.mergeWith(function (a, b) {
  return _ramda2.default.isArrayLike(a) && _ramda2.default.isArrayLike(b) ? _ramda2.default.concat(a, b) : typeof a === 'function' && typeof b === 'function' ? (0, _redux.compose)(a, b) : _ramda2.default.merge(a, b);
}), []);

var combineStack = function combineStack(userStack, initial) {
  return mergeStack(_ramda2.default.concat(_ramda2.default.map(function (_ref) {
    var _ref$reducers = _ref.reducers,
        reducers = _ref$reducers === undefined ? [] : _ref$reducers,
        _ref$enhancers = _ref.enhancers,
        enhancers = _ref$enhancers === undefined ? [] : _ref$enhancers,
        _ref$composers = _ref.composers,
        composers = _ref$composers === undefined ? [] : _ref$composers;
    return { reducers: reducers, enhancers: enhancers, composers: composers };
  }, userStack), [initial]));
};

var buildStack = function buildStack() {
  var userStack = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var initial = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : { reducers: {}, enhancers: [], composers: [_redux.compose] };

  var combinedStack = combineStack(userStack, initial);
  var stackCompose = combinedStack.composers && combinedStack.composers.length > 0 ? _redux.compose.apply(undefined, _toConsumableArray(combinedStack.composers)) : _redux.compose; // inception!
  var stack = { reducers: combinedStack.reducers, enhancer: stackCompose.apply(undefined, _toConsumableArray(combinedStack.enhancers)), meta: userStack };
  return stack;
};

exports.buildStack = buildStack;