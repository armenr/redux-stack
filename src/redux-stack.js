import R from 'ramda'
import { compose } from 'redux'

// TODO: stack formatter + printer dump at build time, etc.

// composition for arrays, hashes, and functions
const mergeStack = R.reduce(R.mergeWith((a, b) =>
    R.isArrayLike(a) && R.isArrayLike(b) ?
      R.concat(a, b)
      : (typeof a === 'function' && typeof b === 'function' ? compose(a, b) : R.merge(a, b))), [])

const combineStack = (userStack, initial) => mergeStack(R.concat(
    R.map(({reducers = [], enhancers = [], composers = []}) => ({reducers, enhancers, composers}), userStack),
    [initial]
))

const buildStack = (userStack = [], initial = { reducers: {}, enhancers: [], composers: [compose] }) => {
  const combinedStack = combineStack(userStack, initial)
  const stackCompose = combinedStack.composers && combinedStack.composers.length > 0 ? compose(...combinedStack.composers) : compose // inception!
  const stack = { reducers: combinedStack.reducers, enhancer: stackCompose(...combinedStack.enhancers), meta: userStack }
  return stack
}

export { buildStack }
