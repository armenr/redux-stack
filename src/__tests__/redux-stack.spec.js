import { buildStack } from '../redux-stack'
describe('buildStack', ()=>{
    it('with empty user stack', ()=>{
      const stack = buildStack()
      expect(stack).toMatchSnapshot()
      expect(stack.enhancer(1)).toEqual(1)
    })
    it('with staggered stack', ()=>{
      const stack = buildStack(
        [
          {reducers: { users: x=>x }},
          {enhancers: [x=>x*2]},
          {composers: [x=>x]},
        ]
      )
      expect(stack).toMatchSnapshot()
      expect(stack.enhancer(1)).toEqual(2)
    })
    it('with additive stack', ()=>{
      const stack = buildStack(
        [
          {reducers: { users: x=>x }, enhancers: [x=>x]},
          {reducers: { posts: x=>x }},
          {enhancers: [x=>x*4]},
          {composers: [x=>x]},
        ]
      )
      expect(stack).toMatchSnapshot()
      expect(stack.enhancer(1)).toEqual(4)
    })
})
