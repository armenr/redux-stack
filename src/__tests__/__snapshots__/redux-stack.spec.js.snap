exports[`buildStack with additive stack 1`] = `
Object {
  "enhancer": [Function],
  "meta": Array [
    Object {
      "enhancers": Array [
        [Function],
      ],
      "reducers": Object {
        "users": [Function],
      },
    },
    Object {
      "reducers": Object {
        "posts": [Function],
      },
    },
    Object {
      "enhancers": Array [
        [Function],
      ],
    },
    Object {
      "composers": Array [
        [Function],
      ],
    },
  ],
  "reducers": Object {
    "posts": [Function],
    "users": [Function],
  },
}
`;

exports[`buildStack with empty user stack 1`] = `
Object {
  "enhancer": [Function],
  "meta": Array [],
  "reducers": Object {},
}
`;

exports[`buildStack with staggered stack 1`] = `
Object {
  "enhancer": [Function],
  "meta": Array [
    Object {
      "reducers": Object {
        "users": [Function],
      },
    },
    Object {
      "enhancers": Array [
        [Function],
      ],
    },
    Object {
      "composers": Array [
        [Function],
      ],
    },
  ],
  "reducers": Object {
    "users": [Function],
  },
}
`;
